utils/types/pattern.js

```diff
@@ -0,0 +1,13 @@
+class PatternMatchable {
+    constructor(n) {
+        this.n = n;
+    }
+
+    static on(ref, args) {
```

Comment:
Maybe something in the name to convey that this enforces exhaustive matching for generic `on` patterns? As opposed to being a handler itself. I can't think of a good one though